docs/src/literate-tutorials/gpu_assembly.jl

```diff
@@ -0,0 +1,323 @@
+# # [Heat equation](@id tutorial-heat-equation)
+#
+# ![](heat_square.png)
+#
+# *Figure 1*: Temperature field on the unit square with an internal uniform heat source
+# solved with homogeneous Dirichlet boundary conditions on the boundary.
+#
+#-
+#md # !!! tip
+#md #     This example is also available as a Jupyter notebook:
+#md #     [`heat_equation.ipynb`](@__NBVIEWER_ROOT_URL__/examples/heat_equation.ipynb).
+#-
+#
+# ## Introduction
+#
+# The heat equation is the "Hello, world!" equation of finite elements.
+# Here we solve the equation on a unit square, with a uniform internal source.
+# The strong form of the (linear) heat equation is given by
+#
+# ```math
+#  -\nabla \cdot (k \nabla u) = f  \quad \textbf{x} \in \Omega,
+# ```
+#
+# where $u$ is the unknown temperature field, $k$ the heat conductivity,
+# $f$ the heat source and $\Omega$ the domain. For simplicity we set $f = 1$
+# and $k = 1$. We will consider homogeneous Dirichlet boundary conditions such that
+# ```math
+# u(\textbf{x}) = 0 \quad \textbf{x} \in \partial \Omega,
+# ```
+# where $\partial \Omega$ denotes the boundary of $\Omega$.
+# The resulting weak form is given given as follows: Find ``u \in \mathbb{U}`` such that
+# ```math
+# \int_{\Omega} \nabla \delta u \cdot \nabla u \ d\Omega = \int_{\Omega} \delta u \ d\Omega \quad \forall \delta u \in \mathbb{T},
+# ```
+# where $\delta u$ is a test function, and where $\mathbb{U}$ and $\mathbb{T}$ are suitable
+# trial and test function sets, respectively.
+#-
+# ## Commented Program
+#
+# Now we solve the problem in Ferrite. What follows is a program spliced with comments.
+#md # The full program, without comments, can be found in the next [section](@ref heat_equation-plain-program).
+#
+# First we load Ferrite, and some other packages we need
+using Ferrite, CUDA
+using IterativeSolvers, LinearAlgebra
+
+### TODO Extension
+import Adapt
+using StaticArrays
+struct GPUQuadratureRule{N,T,dim}
+    weights::SVector{N,T}
+    points::SVector{N,Vec{dim,T}}
+end
+function Adapt.adapt_structure(to, qr::QuadratureRule{shape,T,dim}) where {shape,T,dim}
+    N = length(qr.weights)
+    GPUQuadratureRule{N,T,dim}(SVector{N,T}(qr.weights), SVector{N,Vec{dim,T}}(qr.points))
+end
+function Adapt.adapt_structure(to, nodes::Vector{Node})
+    CuArray(get_node_coordinate.(nodes))
+end
+### TODO Adapt dofhandler
+
+# TODO not sure how to do this automatically
```

Comment:
Driveby comment: I think we could change https://github.com/Ferrite-FEM/Ferrite.jl/blob/3cff6011bcf05797b9759a157aabc28fc0bc6ee3/src/interpolations.jl#L521

to `@boundscheck throw(ArgumentError("no shape function $i for interpolation $ip"))`

(also in all other places) and then in this example use `@inbounds`?

Reply:
Well, this is rather hacky. Give me a bit more time to think about this.

Reply:
Right, this is actually not type stable. 
Probably should return `eltype(ξ)(NaN)` after the `@boundscheck` and update the operations to not multiply by floats. 

But I wouldn't call that hacky though since it just allows `@inbounds` to be used to elude the bounds check?